CBinding/CMake/CMakeFileFormat.cs

```diff
@@ -0,0 +1,289 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Collections.Generic;
+
+using MonoDevelop.Core;
+using MonoDevelop.Projects;
+using MonoDevelop.Ide.Tasks;
+using MonoDevelop.Core.Logging;
+
+namespace CBinding {
+	public class CMakeFileFormat
+	{
+		//File path of the current CMakeLists.txt
+		public FilePath File;
+		public CMakeProject Project;
+		public string ProjectName = ""; //Comes from `project (project_name)` cmake command.
+		public Dictionary<string, CMakeCommand> SetCommands = new Dictionary<string, CMakeCommand> ();
+		public Dictionary<string, CMakeTarget> Targets = new Dictionary<string, CMakeTarget> ();
+		public Dictionary<FilePath, CMakeFileFormat> Children = new Dictionary<FilePath, CMakeFileFormat> ();
+		
+		public CMakeFileFormat Parent;
+		Dictionary<string, CMakeCommand> allCommands = new Dictionary<string, CMakeCommand> ();
+		Stack<string> blocks = new Stack<string> ();
+		MonoDevelop.Projects.Text.TextFile contentFile;
+		
+		//[A-Za-z_][A-Za-z0-9_]*	: ID
+		//\[=[0-9]*\[.*?\]=[0-9]*\]	: bracket argument.
+		string commandPattern = @"[A-Za-z_][A-Za-z0-9_]*\s*\((\"".*?\""|\[=[0-9]*\[.*?\]=[0-9]*\]|.)*?\)";
+		string blockCommentPattern = @"#\[\[.*?\]\]";
+		string lineCommentPattern = @"#.*?\n";
+
+		bool checkSyntax ()
+		{
+			var checkText = Regex.Replace (contentFile.Text, lineCommentPattern, "\n", RegexOptions.Singleline);
+			checkText = Regex.Replace (checkText, blockCommentPattern, "", RegexOptions.Singleline);
+			checkText = Regex.Replace (checkText, @"\s+", " ", RegexOptions.Singleline);
+			checkText = Regex.Replace (checkText, commandPattern, "", RegexOptions.Singleline);
+			if (checkText.Trim ().Length > 0) {
+				MonoDevelop.Ide.MessageService.ShowError ("Couldn't the load project.", String.Format ("Syntax errors found in {0}", File.ToString ()));
+				throw new Exception (String.Format ("Syntax errors found in {0}", File.ToString ()));
+			}
+			return true;
+		}
+
+		IEnumerable<Match> readNextCommand ()
+		{
+			if (checkSyntax ()) {
+				string fullPattern = commandPattern + '|' + lineCommentPattern + '|' + blockCommentPattern;
+				foreach (Match m in Regex.Matches (contentFile.Text, fullPattern, RegexOptions.Singleline)) {
+					if (m.Value.StartsWith ("#"))
+						continue; //Skip comments.
+					yield return m;
+				}
+            }
+		}
+
+		void parseCommand (Match commandMatch)
+		{
+			string commandName = Regex.Match (commandMatch.Value, @"^[A-Za-z_][A-Za-z0-9_]*").Value;
+			CMakeCommand command = new CMakeCommand (commandName, commandMatch, this);
```

Comment:
This is an idea place to use `var` because it's obvious what the type is:

    var command = new CMakeCommand (commandName, commandMatch, this);

I would advise against using var for method return values though as it's not so obvious - e.g. `var foo = DoStuff()` - what type is foo?

If you don't have source analysis turned on I recommend you try it as it'll suggest improvements like this.